src/lxc/lxccontainer.h

```diff
@@ -1101,6 +1101,7 @@ struct lxc_log {
 	const char *file;
 	const char *level;
 	const char *prefix;
+	bool default_level;
```

Comment:
Hm, not sure if we need an API extension for this. I would like to avoid this. Wouldn't it suffice if you did:
```diff
diff --git a/src/lxc/tools/lxc_attach.c b/src/lxc/tools/lxc_attach.c
index d33d9c44..28fb6905 100644
--- a/src/lxc/tools/lxc_attach.c
+++ b/src/lxc/tools/lxc_attach.c
@@ -211,6 +211,7 @@ Options :\n\
 	.options  = my_longopts,
 	.parser   = my_parser,
 	.checker  = NULL,
+	.log_priority = "ERROR",
 };
 
 static bool stdfd_is_pty(void)
diff --git a/src/lxc/tools/lxc_autostart.c b/src/lxc/tools/lxc_autostart.c
index 5d044525..7dd8cc58 100644
--- a/src/lxc/tools/lxc_autostart.c
+++ b/src/lxc/tools/lxc_autostart.c
@@ -102,6 +102,7 @@ Options:\n\
 	.parser   = my_parser,
 	.checker  = NULL,
 	.timeout = 60,
+	.log_priority = "ERROR",
 };
 
 static int list_contains_entry(char *str_ptr, struct lxc_list *p1) {
diff --git a/src/lxc/tools/lxc_cgroup.c b/src/lxc/tools/lxc_cgroup.c
index 06706c84..69f8ae19 100644
--- a/src/lxc/tools/lxc_cgroup.c
+++ b/src/lxc/tools/lxc_cgroup.c
@@ -64,6 +64,7 @@ Options :\n\
 	.options  = my_longopts,
 	.parser   = NULL,
 	.checker  = my_checker,
+	.log_priority = "ERROR",
 };
 
 int main(int argc, char *argv[])
diff --git a/src/lxc/tools/lxc_checkpoint.c b/src/lxc/tools/lxc_checkpoint.c
index 5bdc523c..2f196f34 100644
--- a/src/lxc/tools/lxc_checkpoint.c
+++ b/src/lxc/tools/lxc_checkpoint.c
@@ -153,6 +153,7 @@ Options :\n\
 	.parser    = my_parser,
 	.daemonize = 1,
 	.checker   = my_checker,
+	.log_priority = "ERROR",
 };
 
 static bool checkpoint(struct lxc_container *c)
diff --git a/src/lxc/tools/lxc_console.c b/src/lxc/tools/lxc_console.c
index 11cddd5f..e10aa9fb 100644
--- a/src/lxc/tools/lxc_console.c
+++ b/src/lxc/tools/lxc_console.c
@@ -90,6 +90,7 @@ Options :\n\
 	.checker  = NULL,
 	.ttynum = -1,
 	.escape = 1,
+	.log_priority = "ERROR",
 };
 
 int main(int argc, char *argv[])
diff --git a/src/lxc/tools/lxc_copy.c b/src/lxc/tools/lxc_copy.c
index 33e4f9a7..3e9757e7 100644
--- a/src/lxc/tools/lxc_copy.c
+++ b/src/lxc/tools/lxc_copy.c
@@ -125,6 +125,7 @@ Options :\n\
 	.daemonize = 1,
 	.quiet = false,
 	.tmpfs = false,
+	.log_priority = "ERROR",
 };
 
 static struct mnts *add_mnt(struct mnts **mnts, unsigned int *num,
diff --git a/src/lxc/tools/lxc_create.c b/src/lxc/tools/lxc_create.c
index f704d174..037e82e4 100644
--- a/src/lxc/tools/lxc_create.c
+++ b/src/lxc/tools/lxc_create.c
@@ -173,6 +173,7 @@ Options :\n\
 	.options  = my_longopts,
 	.parser   = my_parser,
 	.checker  = NULL,
+	.log_priority = "ERROR",
 };
 
 static bool validate_bdev_args(struct lxc_arguments *a)
diff --git a/src/lxc/tools/lxc_destroy.c b/src/lxc/tools/lxc_destroy.c
index 35514689..3c8eaf9f 100644
--- a/src/lxc/tools/lxc_destroy.c
+++ b/src/lxc/tools/lxc_destroy.c
@@ -58,6 +58,7 @@ Options :\n\
 	.parser   = my_parser,
 	.checker  = NULL,
 	.task     = DESTROY,
+	.log_priority = "ERROR",
 };
 
 static bool do_destroy(struct lxc_container *c);
diff --git a/src/lxc/tools/lxc_device.c b/src/lxc/tools/lxc_device.c
index 4ac0187e..0981df8c 100644
--- a/src/lxc/tools/lxc_device.c
+++ b/src/lxc/tools/lxc_device.c
@@ -58,6 +58,7 @@ Options :\n\
 	.options  = my_longopts,
 	.parser   = NULL,
 	.checker  = NULL,
+	.log_priority = "ERROR",
 };
 
 static bool is_interface(const char *dev_name, pid_t pid)
diff --git a/src/lxc/tools/lxc_execute.c b/src/lxc/tools/lxc_execute.c
index 2f79b74a..73ff19d0 100644
--- a/src/lxc/tools/lxc_execute.c
+++ b/src/lxc/tools/lxc_execute.c
@@ -117,6 +117,7 @@ Options :\n\
 	.options  = my_longopts,
 	.parser   = my_parser,
 	.daemonize = 0,
+	.log_priority = "ERROR",
 };
 
 static bool set_argv(struct lxc_container *c, struct lxc_arguments *args)
diff --git a/src/lxc/tools/lxc_freeze.c b/src/lxc/tools/lxc_freeze.c
index 88e1f38b..00793ea2 100644
--- a/src/lxc/tools/lxc_freeze.c
+++ b/src/lxc/tools/lxc_freeze.c
@@ -53,6 +53,7 @@ Options :\n\
 	.options  = my_longopts,
 	.parser   = NULL,
 	.checker  = NULL,
+	.log_priority = "ERROR",
 };
 
 int main(int argc, char *argv[])
diff --git a/src/lxc/tools/lxc_info.c b/src/lxc/tools/lxc_info.c
index 16369b72..7338977e 100644
--- a/src/lxc/tools/lxc_info.c
+++ b/src/lxc/tools/lxc_info.c
@@ -101,6 +101,7 @@ Options :\n\
 	.options  = my_longopts,
 	.parser   = my_parser,
 	.checker  = NULL,
+	.log_priority = "ERROR",
 };
 
 static void str_chomp(char *buf)
diff --git a/src/lxc/tools/lxc_ls.c b/src/lxc/tools/lxc_ls.c
index 32278d49..591f7fe5 100644
--- a/src/lxc/tools/lxc_ls.c
+++ b/src/lxc/tools/lxc_ls.c
@@ -201,6 +201,7 @@ Options :\n\
 	.options = my_longopts,
 	.parser = my_parser,
 	.ls_nesting = 0,
+	.log_priority = "ERROR",
 };
 
 int main(int argc, char *argv[])
diff --git a/src/lxc/tools/lxc_monitor.c b/src/lxc/tools/lxc_monitor.c
index abeccdee..e918c1f7 100644
--- a/src/lxc/tools/lxc_monitor.c
+++ b/src/lxc/tools/lxc_monitor.c
@@ -89,6 +89,7 @@ Options :\n\
 	.parser   = my_parser,
 	.checker  = NULL,
 	.lxcpath_additional = -1,
+	.log_priority = "ERROR",
 };
 
 static void close_fds(struct pollfd *fds, nfds_t nfds)
diff --git a/src/lxc/tools/lxc_snapshot.c b/src/lxc/tools/lxc_snapshot.c
index f943bdd1..e56dea20 100644
--- a/src/lxc/tools/lxc_snapshot.c
+++ b/src/lxc/tools/lxc_snapshot.c
@@ -67,6 +67,7 @@ Options :\n\
 	.parser = my_parser,
 	.checker = NULL,
 	.task = SNAP,
+	.log_priority = "ERROR",
 };
 
 static int do_snapshot(struct lxc_container *c, char *commentfile);
diff --git a/src/lxc/tools/lxc_start.c b/src/lxc/tools/lxc_start.c
index 398b5634..75eb2646 100644
--- a/src/lxc/tools/lxc_start.c
+++ b/src/lxc/tools/lxc_start.c
@@ -166,6 +166,7 @@ Options :\n\
 	.checker   = NULL,
 	.daemonize = 1,
 	.pidfile = NULL,
+	.log_priority = "ERROR",
 };
 
 int main(int argc, char *argv[])
diff --git a/src/lxc/tools/lxc_stop.c b/src/lxc/tools/lxc_stop.c
index c1b6a476..01596c2d 100644
--- a/src/lxc/tools/lxc_stop.c
+++ b/src/lxc/tools/lxc_stop.c
@@ -95,6 +95,7 @@ Options :\n\
 	.parser   = my_parser,
 	.checker  = NULL,
 	.timeout  = -2,
+	.log_priority = "ERROR",
 };
 
 int main(int argc, char *argv[])
diff --git a/src/lxc/tools/lxc_top.c b/src/lxc/tools/lxc_top.c
index b873be64..f7098809 100644
--- a/src/lxc/tools/lxc_top.c
+++ b/src/lxc/tools/lxc_top.c
@@ -137,6 +137,7 @@ Options :\n\
 	.parser   = my_parser,
 	.checker  = NULL,
 	.lxcpath_additional = -1,
+	.log_priority = "ERROR",
 };
 
 static void stdin_tios_restore(void)
diff --git a/src/lxc/tools/lxc_unfreeze.c b/src/lxc/tools/lxc_unfreeze.c
index a0b56900..6bf71f3f 100644
--- a/src/lxc/tools/lxc_unfreeze.c
+++ b/src/lxc/tools/lxc_unfreeze.c
@@ -52,6 +52,7 @@ Options :\n\
 	.options  = my_longopts,
 	.parser   = NULL,
 	.checker  = NULL,
+	.log_priority = "ERROR",
 };
 
 int main(int argc, char *argv[])
diff --git a/src/lxc/tools/lxc_unshare.c b/src/lxc/tools/lxc_unshare.c
index 720ecb37..6da34cf4 100644
--- a/src/lxc/tools/lxc_unshare.c
+++ b/src/lxc/tools/lxc_unshare.c
@@ -104,6 +104,7 @@ Options :\n\
 	.checker      = NULL,
 	.daemonize    = 0,
 	.pidfile      = NULL,
+	.log_priority = "ERROR",
 };
 
 static int my_parser(struct lxc_arguments *args, int c, char *arg)
diff --git a/src/lxc/tools/lxc_wait.c b/src/lxc/tools/lxc_wait.c
index deecc773..2f429115 100644
--- a/src/lxc/tools/lxc_wait.c
+++ b/src/lxc/tools/lxc_wait.c
@@ -81,6 +81,7 @@ Options :\n\
 	.parser   = my_parser,
 	.checker  = my_checker,
 	.timeout = -1,
+	.log_priority = "ERROR",
 };
 
 int main(int argc, char *argv[])
```

Reply:
@brauner Yes, I know. But it is so complicated. If default priority is set to "struct lxc_arguments" you suggest, a log file is always created when tools are launched. So, it causes another behavior as before. 

Current logic is,
1. user does not specify priority or log file => stderr and log file is not working
2. user specifies priority or log file => stderr and log file is working

My proposal is that,,

Although user doe snot specify priority or log file explicitly, tools print stderr logs of error level without log file.

Please see lxc-unshare included first commit.

Do you know why lxc_log is exposed to external header(lxcconatiner.h)?

Thanks. :)

Reply:
> Yes, I know. But it is so complicated. If default priority is set to "struct lxc_arguments" you suggest, a log file is always created when tools are launched. So, it causes another behavior as before.

Hm, looking at the code:
```diff --git a/src/lxc/tools/lxc_attach.c b/src/lxc/tools/lxc_attach.c
index d33d9c44..6eec0f53 100644
--- a/src/lxc/tools/lxc_attach.c
+++ b/src/lxc/tools/lxc_attach.c
@@ -211,6 +211,8 @@ Options :\n\
 	.options  = my_longopts,
 	.parser   = my_parser,
 	.checker  = NULL,
+	.log_priority = "ERROR",
+	.log_file = "none",
 };
 
 static bool stdfd_is_pty(void)
diff --git a/src/lxc/tools/lxc_autostart.c b/src/lxc/tools/lxc_autostart.c
index 5d044525..6b321f2c 100644
--- a/src/lxc/tools/lxc_autostart.c
+++ b/src/lxc/tools/lxc_autostart.c
@@ -102,6 +102,8 @@ Options:\n\
 	.parser   = my_parser,
 	.checker  = NULL,
 	.timeout = 60,
+	.log_priority = "ERROR",
+	.log_file = "none",
 };
 
 static int list_contains_entry(char *str_ptr, struct lxc_list *p1) {
diff --git a/src/lxc/tools/lxc_cgroup.c b/src/lxc/tools/lxc_cgroup.c
index 06706c84..2e4ccb9d 100644
--- a/src/lxc/tools/lxc_cgroup.c
+++ b/src/lxc/tools/lxc_cgroup.c
@@ -64,6 +64,8 @@ Options :\n\
 	.options  = my_longopts,
 	.parser   = NULL,
 	.checker  = my_checker,
+	.log_priority = "ERROR",
+	.log_file = "none",
 };
 
 int main(int argc, char *argv[])
diff --git a/src/lxc/tools/lxc_checkpoint.c b/src/lxc/tools/lxc_checkpoint.c
index 5bdc523c..62df33e7 100644
--- a/src/lxc/tools/lxc_checkpoint.c
+++ b/src/lxc/tools/lxc_checkpoint.c
@@ -153,6 +153,8 @@ Options :\n\
 	.parser    = my_parser,
 	.daemonize = 1,
 	.checker   = my_checker,
+	.log_priority = "ERROR",
+	.log_file = "none",
 };
 
 static bool checkpoint(struct lxc_container *c)
diff --git a/src/lxc/tools/lxc_console.c b/src/lxc/tools/lxc_console.c
index 11cddd5f..a02b83fe 100644
--- a/src/lxc/tools/lxc_console.c
+++ b/src/lxc/tools/lxc_console.c
@@ -90,6 +90,8 @@ Options :\n\
 	.checker  = NULL,
 	.ttynum = -1,
 	.escape = 1,
+	.log_priority = "ERROR",
+	.log_file = "none",
 };
 
 int main(int argc, char *argv[])
diff --git a/src/lxc/tools/lxc_copy.c b/src/lxc/tools/lxc_copy.c
index 33e4f9a7..6519edcf 100644
--- a/src/lxc/tools/lxc_copy.c
+++ b/src/lxc/tools/lxc_copy.c
@@ -125,6 +125,8 @@ Options :\n\
 	.daemonize = 1,
 	.quiet = false,
 	.tmpfs = false,
+	.log_priority = "ERROR",
+	.log_file = "none",
 };
 
 static struct mnts *add_mnt(struct mnts **mnts, unsigned int *num,
diff --git a/src/lxc/tools/lxc_create.c b/src/lxc/tools/lxc_create.c
index f704d174..6336dac2 100644
--- a/src/lxc/tools/lxc_create.c
+++ b/src/lxc/tools/lxc_create.c
@@ -173,6 +173,8 @@ Options :\n\
 	.options  = my_longopts,
 	.parser   = my_parser,
 	.checker  = NULL,
+	.log_priority = "ERROR",
+	.log_file = "none",
 };
 
 static bool validate_bdev_args(struct lxc_arguments *a)
diff --git a/src/lxc/tools/lxc_destroy.c b/src/lxc/tools/lxc_destroy.c
index 35514689..da4d1a3d 100644
--- a/src/lxc/tools/lxc_destroy.c
+++ b/src/lxc/tools/lxc_destroy.c
@@ -58,6 +58,8 @@ Options :\n\
 	.parser   = my_parser,
 	.checker  = NULL,
 	.task     = DESTROY,
+	.log_priority = "ERROR",
+	.log_file = "none",
 };
 
 static bool do_destroy(struct lxc_container *c);
diff --git a/src/lxc/tools/lxc_device.c b/src/lxc/tools/lxc_device.c
index 4ac0187e..8a7fe735 100644
--- a/src/lxc/tools/lxc_device.c
+++ b/src/lxc/tools/lxc_device.c
@@ -58,6 +58,8 @@ Options :\n\
 	.options  = my_longopts,
 	.parser   = NULL,
 	.checker  = NULL,
+	.log_priority = "ERROR",
+	.log_file = "none",
 };
 
 static bool is_interface(const char *dev_name, pid_t pid)
diff --git a/src/lxc/tools/lxc_execute.c b/src/lxc/tools/lxc_execute.c
index 2f79b74a..b40c76f3 100644
--- a/src/lxc/tools/lxc_execute.c
+++ b/src/lxc/tools/lxc_execute.c
@@ -117,6 +117,8 @@ Options :\n\
 	.options  = my_longopts,
 	.parser   = my_parser,
 	.daemonize = 0,
+	.log_priority = "ERROR",
+	.log_file = "none",
 };
 
 static bool set_argv(struct lxc_container *c, struct lxc_arguments *args)
diff --git a/src/lxc/tools/lxc_freeze.c b/src/lxc/tools/lxc_freeze.c
index 88e1f38b..c8025ed6 100644
--- a/src/lxc/tools/lxc_freeze.c
+++ b/src/lxc/tools/lxc_freeze.c
@@ -53,6 +53,8 @@ Options :\n\
 	.options  = my_longopts,
 	.parser   = NULL,
 	.checker  = NULL,
+	.log_priority = "ERROR",
+	.log_file = "none",
 };
 
 int main(int argc, char *argv[])
diff --git a/src/lxc/tools/lxc_info.c b/src/lxc/tools/lxc_info.c
index 16369b72..879042fa 100644
--- a/src/lxc/tools/lxc_info.c
+++ b/src/lxc/tools/lxc_info.c
@@ -101,6 +101,8 @@ Options :\n\
 	.options  = my_longopts,
 	.parser   = my_parser,
 	.checker  = NULL,
+	.log_priority = "ERROR",
+	.log_file = "none",
 };
 
 static void str_chomp(char *buf)
diff --git a/src/lxc/tools/lxc_ls.c b/src/lxc/tools/lxc_ls.c
index 32278d49..a09a0b2e 100644
--- a/src/lxc/tools/lxc_ls.c
+++ b/src/lxc/tools/lxc_ls.c
@@ -201,6 +201,8 @@ Options :\n\
 	.options = my_longopts,
 	.parser = my_parser,
 	.ls_nesting = 0,
+	.log_priority = "ERROR",
+	.log_file = "none",
 };
 
 int main(int argc, char *argv[])
diff --git a/src/lxc/tools/lxc_monitor.c b/src/lxc/tools/lxc_monitor.c
index abeccdee..df992927 100644
--- a/src/lxc/tools/lxc_monitor.c
+++ b/src/lxc/tools/lxc_monitor.c
@@ -89,6 +89,8 @@ Options :\n\
 	.parser   = my_parser,
 	.checker  = NULL,
 	.lxcpath_additional = -1,
+	.log_priority = "ERROR",
+	.log_file = "none",
 };
 
 static void close_fds(struct pollfd *fds, nfds_t nfds)
diff --git a/src/lxc/tools/lxc_snapshot.c b/src/lxc/tools/lxc_snapshot.c
index f943bdd1..cf41d327 100644
--- a/src/lxc/tools/lxc_snapshot.c
+++ b/src/lxc/tools/lxc_snapshot.c
@@ -67,6 +67,8 @@ Options :\n\
 	.parser = my_parser,
 	.checker = NULL,
 	.task = SNAP,
+	.log_priority = "ERROR",
+	.log_file = "none",
 };
 
 static int do_snapshot(struct lxc_container *c, char *commentfile);
diff --git a/src/lxc/tools/lxc_start.c b/src/lxc/tools/lxc_start.c
index 398b5634..abbdd718 100644
--- a/src/lxc/tools/lxc_start.c
+++ b/src/lxc/tools/lxc_start.c
@@ -166,6 +166,8 @@ Options :\n\
 	.checker   = NULL,
 	.daemonize = 1,
 	.pidfile = NULL,
+	.log_priority = "ERROR",
+	.log_file = "none",
 };
 
 int main(int argc, char *argv[])
diff --git a/src/lxc/tools/lxc_stop.c b/src/lxc/tools/lxc_stop.c
index c1b6a476..552b87de 100644
--- a/src/lxc/tools/lxc_stop.c
+++ b/src/lxc/tools/lxc_stop.c
@@ -95,6 +95,8 @@ Options :\n\
 	.parser   = my_parser,
 	.checker  = NULL,
 	.timeout  = -2,
+	.log_priority = "ERROR",
+	.log_file = "none",
 };
 
 int main(int argc, char *argv[])
diff --git a/src/lxc/tools/lxc_top.c b/src/lxc/tools/lxc_top.c
index b873be64..9f57c6ee 100644
--- a/src/lxc/tools/lxc_top.c
+++ b/src/lxc/tools/lxc_top.c
@@ -137,6 +137,8 @@ Options :\n\
 	.parser   = my_parser,
 	.checker  = NULL,
 	.lxcpath_additional = -1,
+	.log_priority = "ERROR",
+	.log_file = "none",
 };
 
 static void stdin_tios_restore(void)
diff --git a/src/lxc/tools/lxc_unfreeze.c b/src/lxc/tools/lxc_unfreeze.c
index a0b56900..54aaf7ba 100644
--- a/src/lxc/tools/lxc_unfreeze.c
+++ b/src/lxc/tools/lxc_unfreeze.c
@@ -52,6 +52,8 @@ Options :\n\
 	.options  = my_longopts,
 	.parser   = NULL,
 	.checker  = NULL,
+	.log_priority = "ERROR",
+	.log_file = "none",
 };
 
 int main(int argc, char *argv[])
diff --git a/src/lxc/tools/lxc_unshare.c b/src/lxc/tools/lxc_unshare.c
index 720ecb37..18bc0337 100644
--- a/src/lxc/tools/lxc_unshare.c
+++ b/src/lxc/tools/lxc_unshare.c
@@ -104,6 +104,8 @@ Options :\n\
 	.checker      = NULL,
 	.daemonize    = 0,
 	.pidfile      = NULL,
+	.log_priority = "ERROR",
+	.log_file = "none",
 };
 
 static int my_parser(struct lxc_arguments *args, int c, char *arg)
diff --git a/src/lxc/tools/lxc_wait.c b/src/lxc/tools/lxc_wait.c
index deecc773..ba19e8a5 100644
--- a/src/lxc/tools/lxc_wait.c
+++ b/src/lxc/tools/lxc_wait.c
@@ -81,6 +81,8 @@ Options :\n\
 	.parser   = my_parser,
 	.checker  = my_checker,
 	.timeout = -1,
+	.log_priority = "ERROR",
+	.log_file = "none",
 };
 
 int main(int argc, char *argv[])
```

Reply:
@brauner ok :)